src/Feldspar/Vector.hs

```diff
@@ -221,8 +221,11 @@ instance CollMap (Pull sh a) (Pull sh b) where
 
 -- | Store a vector to memory.
 fromPull :: (Type a, Shapely sh) => DPull sh a -> Data [a]
-fromPull vec = parallel (size ext) (\ix -> vec !: fromIndex ext ix)
+fromPull vec = materialize (size ext) $ toLoops (\ ix -> write (toIndex ext ix) (vec ! ix)) ext
   where ext = extent vec
+        toLoops :: Type a => (Shape sh -> Data (Elements a)) -> Shape sh -> Data (Elements a)
+        toLoops f Z = f Z
+        toLoops f (bnds :. n) = toLoops (\ ix -> parFor n $ \ i -> f (ix :. i)) bnds
```

Comment:
You could call `fromVector'` directly. It does the same thing

Reply:
You're right, they are identical, so I did the same transformation for fromVector' and changed fromList to use elements as well.